src/arkode/arkode_relaxation.c

```diff
@@ -0,0 +1,844 @@
+{
+  int retval;
+  ARKodeMem ark_mem;
+  ARKodeRelaxMem relax_mem;
+
+  retval = arkRelaxAccessMem(arkode_mem, "arkRelaxGetNumRelaxFails", &ark_mem,
+                             &relax_mem);
+  if (retval) return retval;
+
+  *relax_fails = relax_mem->num_fails;
+
+  return ARK_SUCCESS;
+}
+
+int arkRelaxGetNumRelaxSolveFails(void* arkode_mem, long int* fails)
+{
+  int retval;
+  ARKodeMem ark_mem;
+  ARKodeRelaxMem relax_mem;
+
+  retval = arkRelaxAccessMem(arkode_mem, "arkRelaxGetNumRelaxSolveFails",
+                             &ark_mem, &relax_mem);
+  if (retval) return retval;
+
+  *fails = relax_mem->nls_fails;
+
+  return ARK_SUCCESS;
+}
+
+int arkRelaxGetNumRelaxSolveIters(void* arkode_mem, long int* iters)
+{
+  int retval;
+  ARKodeMem ark_mem;
+  ARKodeRelaxMem relax_mem;
+
+  retval = arkRelaxAccessMem(arkode_mem, "arkRelaxGetNumRelaxSolveIters",
+                             &ark_mem, &relax_mem);
+  if (retval) return retval;
+
+  *iters = relax_mem->nls_iters;
+
+  return ARK_SUCCESS;
+}
+
+int arkRelaxPrintAllStats(void* arkode_mem, FILE* outfile, SUNOutputFormat fmt)
+{
+  int retval;
+  ARKodeMem ark_mem;
+  ARKodeRelaxMem relax_mem;
+
+  retval = arkRelaxAccessMem(arkode_mem, "arkRelaxPrintAllStats", &ark_mem,
+                             &relax_mem);
+  if (retval) return retval;
+
+  switch(fmt)
+  {
+  case SUN_OUTPUTFORMAT_TABLE:
+    fprintf(outfile, "Relax fn evals               = %ld\n",
+            relax_mem->num_relax_fn_evals);
+    fprintf(outfile, "Relax Jac evals              = %ld\n",
+            relax_mem->num_relax_jac_evals);
+    fprintf(outfile, "Relax fails                  = %ld\n",
+            relax_mem->num_fails);
+    fprintf(outfile, "Relax NLS iters              = %ld\n",
+            relax_mem->nls_iters);
+    fprintf(outfile, "Relax NLS fails              = %ld\n",
+            relax_mem->nls_fails);
+    break;
+  case SUN_OUTPUTFORMAT_CSV:
+    fprintf(outfile, ",Relax fn evals,%ld", relax_mem->num_relax_fn_evals);
+    fprintf(outfile, ",Relax Jac evals,%ld", relax_mem->num_relax_jac_evals);
+    fprintf(outfile, ",Relax fails,%ld", relax_mem->num_fails);
+    fprintf(outfile, ",Relax NLS iters,%ld", relax_mem->nls_iters);
+    fprintf(outfile, ",Relax NLS fails,%ld", relax_mem->nls_fails);
+    break;
+  default:
+    arkProcessError(ark_mem, ARK_ILL_INPUT, "ARKODE", "arkRelaxPrintAllStats",
+                    "Invalid formatting option.");
+    return ARK_ILL_INPUT;
+  }
+
+  return ARK_SUCCESS;
+}
+
+/* =============================================================================
+ * Driver and Stepper Functions
+ * ===========================================================================*/
+
+/* Constructor called by stepper */
+int arkRelaxCreate(void* arkode_mem, ARKRelaxFn relax_fn,
+                   ARKRelaxJacFn relax_jac_fn, ARKRelaxDeltaYFn delta_y_fn,
+                   ARKRelaxDeltaEFn delta_e_fn, ARKRelaxGetOrderFn get_order_fn)
```

Comment:
Why does this require a whole new function to return the method order (instead of just an integer)?  Is this forward-thinking for something like CVODE, where the order can change dynamically, or is it just here because this constructor might be called before the user sets the desired method order or Butcher table?

Reply:
This was meant to be forward-looking with the idea of generalizing this relaxation module to work with other integrator (in particular CVODE).

Reply:
It will also "play nice" with some experiments I'm planning for student projects this Fall (variable step/order Runge--Kutta methods), that may eventually make it into ARKODE.